src/datasource/gateway.js

```diff
@@ -0,0 +1,243 @@
+    }
+
+    throw error;
+
+  }
+
+  perform = async (requestName, ...args) => {
+    const reqConfig = typeof this.requests[requestName] === 'function' ?
+      this.requests[requestName](...args) :
+      this.requests[requestName];
+
+    const response = await this
+      .provider
+      .http
+      .request(reqConfig)
+      .then(this.parseSuccessfulResponse)
+      .catch(this.parseErrorResponse);
+
+    return response;
+  }
+  
+  // does not really need to be bound to `this`, but keeping it for semantics' sake.
+  // arguments explicit forwardong for clear signature lookup, avoid using `...args` 
+  _getBalance = (address, height) => this.perform.call(this, 'getBalance', address, height) 
+  _getTransaction = (id) => this.perform.call(this, 'getTransaction', id)
+  _getApp = (address, height) => this.perform.call(this, 'getApp', address, height)
+  _getNode = (address, height) => this.perform.call(this, 'getNode', address, height)
+  _getAccountTxs = (address, received, prove, page, per_page) => this.perform.call(this, 'getAccountTxs', address, received, prove, page, per_page)
+  _sendRawTx = (fromAddress, tx) => this.perform.call(this, 'sendRawTx', fromAddress, tx)
+
+  // For semantic separation, and for "ease of middlewaring" when necessary.
+  // hook your processors to your cals in here
+  query = {
+    getBalance: this._getBalance,
+    getTransaction: this._getTransaction,
+    getApp: this._getApp,
+    getNode: this._getNode,
+    getAccountTxs: async (...args) => {
+      const rawResponse = await this._getAccountTxs(...args);
+      const response = this.processors.accountTxs.processResponse(rawResponse);
+
+      return response;
+    },
+    sendRawTx: async (fromAddress, tx) => {
+      const request = this.processors.rawTx.processRequest({ fromAddress, tx });
+      const rawResponse = await this._sendRawTx(request.addressHex, request.rawTxBytes);
+      const response = this.processors.rawTx.processResponse(rawResponse);
+
+      return response;
+    }
+  }
+
+  // request/response processors
+  processors = {
+    rawTx: {
+      processRequest: ({ fromAddress, tx }) => ({
+        addressHex: fromAddress.toString('hex'),
+        rawTxBytes: tx.toString('hex'),
+      }),
+      processResponse: (response) => response,
+    },
+
+    accountTxs: {
+      processResponse: (response) => {
+        const base64ToStr = (v) => Buffer.from(v, "base64").toString();
+        const kvToStr = (kvObj) => ({
+          key: base64ToStr(kvObj.key),
+          value: base64ToStr(kvObj.value),
+        });
+
+        const mapEvents = (events) => events ? events.map(
+          (e) => ({ ...e, attributes: e.attributes.map(kvToStr) })
+        ) : []
+
+        const txs = response.txs.map(
+          (tx) => ({
+            ...tx,
+            tx_result: {
+              ...tx.tx_result,
+              events: mapEvents(tx.tx_result.events),
+            }
+          })
+        )
+
+        return { ...response, txs }
```

Comment:
Are we sure we wanna destructure `response`? Thinking we might not know what's inside of this object at all times.

Reply:
We sorta do that's why